storage/migrations/01_oasis_3_consensus.up.sql

```diff
@@ -6,41 +6,51 @@ BEGIN;
 -- Create Damask Upgrade Schema with `chain-id`.
 CREATE SCHEMA IF NOT EXISTS oasis_3;
 
--- Block Data
+-- Custom types - basic
+CREATE DOMAIN uint_numeric NUMERIC(1000,0) CHECK(VALUE >= 0);
+CREATE DOMAIN uint63 BIGINT CHECK(VALUE >= 0);
+CREATE DOMAIN uint31 INTEGER CHECK(VALUE >= 0);
+CREATE DOMAIN hex64 TEXT CHECK(VALUE ~ '^[0-9a-f]{64}$');
+-- base64(ed25519 public key); from https://github.com/oasisprotocol/oasis-core/blob/f95186e3f15ec64bdd36493cde90be359bd17da8/go/common/crypto/signature/signature.go#L90-L90
+CREATE DOMAIN base64_ed25519_pubkey TEXT CHECK(VALUE ~ '^[A-Za-z0-9+/]{43}=$');
+
+-- Custom types - oasis-specific
+CREATE DOMAIN oasis_addr TEXT CHECK(length(VALUE) = 46 AND VALUE ~ '^oasis1');
 
+-- Block Data
 CREATE TABLE oasis_3.blocks
 (
-  height     BIGINT PRIMARY KEY,
-  block_hash TEXT NOT NULL,
+  height     UINT63 PRIMARY KEY,
+  block_hash HEX64 NOT NULL,
   time       TIMESTAMP WITH TIME ZONE NOT NULL,
 
   -- State Root Info
   namespace TEXT NOT NULL,
-  version   BIGINT NOT NULL,
-  type      TEXT NOT NULL,
-  root_hash TEXT NOT NULL,
+  version   UINT63 NOT NULL,
+  type      TEXT NOT NULL,  -- "invalid" | "state-root" | "io-root"; From https://github.com/oasisprotocol/oasis-core/blob/f95186e3f15ec64bdd36493cde90be359bd17da8/go/storage/mkvs/node/node.go#L68-L68
+  root_hash HEX64 NOT NULL,
 
   beacon     BYTEA,
   metadata   JSON
 );
 
 CREATE TABLE oasis_3.transactions
 (
-  block BIGINT NOT NULL REFERENCES oasis_3.blocks(height),
+  block BIGINT NOT NULL REFERENCES oasis_3.blocks(height) DEFERRABLE INITIALLY DEFERRED,
 
-  txn_hash   TEXT NOT NULL,
-  txn_index  INTEGER,
-  nonce      BIGINT NOT NULL,
-  fee_amount NUMERIC,
-  max_gas    NUMERIC,
+  txn_hash   HEX64 NOT NULL,
+  txn_index  UINT31,
+  nonce      UINT63 NOT NULL,
+  fee_amount UINT_NUMERIC,
+  max_gas    UINT_NUMERIC,
   method     TEXT NOT NULL,
-  sender     TEXT NOT NULL,
+  sender     oasis_addr NOT NULL,
   body       BYTEA,
 
   -- Error Fields
   -- This includes an encoding of no error.
   module  TEXT,
-  code    BIGINT,
+  code    UINT31,  -- From https://github.com/oasisprotocol/oasis-core/blob/f95186e3f15ec64bdd36493cde90be359bd17da8/go/consensus/api/transaction/results/results.go#L20-L20
```

Comment:
The underlying go type (linked in the comment) is uint32, with currently-used constants nowhere near the top of the range. So I figure we might as well do the same thing we do with golang uint64: use `uint(N-1)` in postgres and hope for the best :crossed_fingers:.
It is not unreasonable to do IMO because there's not really a reason for the value to ever be above max(uint31), at least in today's codebase. But this solution does have the potential of a midnight pagerduty call. We would ideally do a pre-check in golang (see PR description, last paragraph).